javaslang/src/main/java/javaslang/collection/BitSet.java

```diff
@@ -0,0 +1,1024 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import javaslang.Function1;
+import javaslang.Tuple;
+import javaslang.Tuple2;
+import javaslang.Tuple3;
+import javaslang.control.Option;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.*;
+import java.util.stream.Collector;
+
+import static javaslang.collection.Comparators.naturalComparator;
+
+/**
+ * An immutable {@code BitSet} implementation.
+ *
+ * @author Ruslan Sennov
+ * @since 2.1.0
+ */
+public interface BitSet<T> extends SortedSet<T> {
+
+    long serialVersionUID = 1L;
+
+    class Builder<T> {
+
+        final static Builder<Integer> DEFAULT = new Builder<>(i -> i, i -> i);
+
+        final Function1<Integer, T> fromInt;
+        final Function1<T, Integer> toInt;
+
+        Builder(Function1<Integer, T> fromInt, Function1<T, Integer> toInt) {
+            this.fromInt = fromInt;
+            this.toInt = toInt;
+        }
+
+        public Collector<T, ArrayList<T>, BitSet<T>> collector() {
+            final BinaryOperator<ArrayList<T>> combiner = (left, right) -> {
+                left.addAll(right);
+                return left;
+            };
+            return Collector.of(ArrayList::new, ArrayList::add, combiner, this::ofAll);
```

Comment:
Maybe we can omit the `ArrayList` because `this::ofAll` is the same as `empty().addAll(...)`:

```java
return Collector.of(this::empty, BitSet::add, ?BitSet::addAll?, this::ofAll)
```

For other collections we have special concat operations that can be used instead of `addAll` I think. E.g. `merge`, `union`, etc. Can we do s.th. similar here for the combiner, too (see above)?